[core] Spill tasks from waiting queue

stephanie-wang opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Take 2 on #14094. The issue is that we do not spill tasks from the waiting queue. Also, object store memory is treated differently from other resources, so the current scheduling policy is not enough. There are a few reasons for this:
1. Object store memory can be shared by multiple tasks that have overlapping dependencies.
2. Object store memory can be used in other ways, e.g., for pinned task outputs.
3. Object store memory requirements are computed from the task argument size, not requested explicitly by the user.

To compensate, this PR checks whether a task's dependencies are being fetched or not, to determine whether there is enough object store memory locally.
```
Previous scheduling policy:
  Resources available locally -> do not spill
  Resources not available locally -> spill from the dispatch queue (tasks with arguments already local)
New scheduling policy:
  Resources available locally and deps are being fetched for waiting tasks -> do not spill
  Resources available locally and deps are not being fetched for waiting tasks -> spill
  Resources not available locally -> spill
```

## Related issue number

Closes #14094.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ericl commented on src/ray/object_manager/object_manager.h:
```diff
@@ -198,6 +199,10 @@ class ObjectManager : public ObjectManagerInterface,
   /// Get the port of the object manager rpc server.
   int GetServerPort() const { return object_manager_server_.GetPort(); }
 
+  bool IsPullRequestInactiveDueToOom(uint64_t pull_request_id) const {
```
```suggestion
  bool NotBeingPulledYet(uint64_t pull_request_id) const {
```

ericl replied:
Actually would suggest inverting this: `PullActive`

ericl commented on src/ray/object_manager/pull_manager.cc:
```diff
@@ -162,6 +162,18 @@ void PullManager::UpdatePullsBasedOnAvailableMemory(size_t num_bytes_available)
+    // more requests in the queue, since we do not serve requests until we have
+    // the object size information.
+    at_memory_capacity_ = false;
+  }
```
We can simplify by not checking this. I think it's fine to spill if we don't have object capacity information yet.

ericl commented:
Main comment is to remove the OOM check condition.

rkooo567 commented on src/ray/raylet/scheduling/cluster_resource_scheduler.h:
```diff
@@ -156,7 +160,7 @@ class ClusterResourceScheduler : public ClusterResourceSchedulerInterface {
   std::string GetBestSchedulableNode(
       const std::unordered_map<std::string, double> &task_request, bool actor_creation,
-      int64_t *violations, bool *is_infeasible);
+      bool force_spillback, int64_t *violations, bool *is_infeasible);
```
Why don't we just make it have a default argument false?

stephanie-wang replied:
I think it's better to be explicit.

stephanie-wang commented:
> Main comment is to remove the OOM check condition.

We need the OOM check to let the scheduling policy know whether we should prefer local scheduling or not. We can sync offline if this is still unclear.

rkooo567 commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
 void ClusterTaskManager::ScheduleAndDispatchTasks() {
   SchedulePendingTasks();
   DispatchScheduledTasksToWorkers(worker_pool_, leased_workers_);
+  // TODO(swang): Spill from waiting queue first? Otherwise, we may end up
```
This sounds like a good idea? Why do you not do that here? Any edge case in your mind?

stephanie-wang replied:
It's a bit hard to reason about. I'd rather do this if there is actually a bug associated with it.

rkooo567 commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
+void ClusterTaskManager::SpillWaitingTasks() {
+  // Try to spill waiting tasks to a remote node, prioritizing those at the end
+  // of the queue. Waiting tasks are spilled if there are enough remote
+  // resources AND (we have no resources available locally or we cannot fetch
```
```suggestion
  // resources AND (we have no resources available locally OR we cannot fetch
```

rkooo567 commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
+    int64_t _unused;
+    bool is_infeasible;
+    // TODO(swang): The policy currently does not account for object store
+    // memory availability. Ideally, we should pick the node with the most
```
Isn't it causing flooding? I think it'd be okay to just do random spilling like now

rkooo567 commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
+  // due to lack of remote resources, it is possible that a waiting task that
+  // is earlier in the queue could have been scheduled to a remote node.
+  auto it = waiting_task_queue_.end();
+  while (it != waiting_task_queue_.begin()) {
```
Can we specify the end of the queue is not the same as pulling order?

rkooo567 commented on src/ray/raylet/scheduling/cluster_task_manager_test.cc:
```diff
@@ -954,6 +960,94 @@ TEST_F(ClusterTaskManagerTest, RleaseAndReturnWorkerCpuResources) {
+    dependency_manager_.blocked_tasks.insert(task.GetTaskSpecification().TaskId());
+  }
+  task_manager_.ScheduleAndDispatchTasks();
+  ASSERT_EQ(num_callbacks, 2);
```
Nice test!

rkooo567 reviewed (approved):
LGTM.

stephanie-wang commented:
Synced with @ericl offline and we decided on these changes:
- check if pull request is active or not instead of checking at memory capacity. Will also add a check to make sure that the pull request has all its object size information, since this is also a reason why it could be inactive.
- add a Python test to make sure we are triggering spillback correctly in the case that there are enough CPUs but not enough memory.
- invoke the scheduler periodically to handle the above case.

stephanie-wang commented:
FYI, I added a regression test for spilling back on OOM but there is a bug in the resource reporting that is actually causing resources to get reported on every heartbeat, so the test is actually passing. I added a TODO to trigger spillback from the waiting queue instead of doing it in this PR.

stephanie-wang commented:
Note: Need to fix tests.

ericl commented on src/ray/raylet/scheduling/cluster_task_manager.cc:
```diff
@@ -887,6 +895,62 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
+    // Check whether this task's dependencies are not being pulled.  If this is
+    // true, then we should force the task onto a remote feasible node, even if
+    // we have enough resources available locally for placement.
+    bool force_spillback = !task_dependency_manager_.TaskDependenciesPending(task_id);
```
```suggestion
    bool force_spillback = task_dependency_manager_.CannotExecuteLocallyYet(task_id);
```

ericl replied:
(avoid double negative)

stephanie-wang replied:
I don't think this name is accurate. The call just returns whether dependencies are being fetched, not whether the dependencies are local yet.

ericl replied:
(also avoid "pending" since arguably everything is "pending" in the task dep manager)

stephanie-wang replied:
How about "TaskDependenciesBlocked" then?

ericl replied:
:+1: blocked

stephanie-wang commented:
FYI, the load-balancing test that I added passes consistently on my laptop but not on Travis. I'm not sure exactly why, but it looks like it probably has something to do with the raylets ping-ponging waiting tasks, since they all think that another node has enough CPU resources to run the task. I'm reverting the test to the previous version that just checks whether it finishes instead of checking for even load. The other test that is added in this PR is good enough as a regression test.

ericl commented:
LINT

rkooo567 commented:
`test_scheduling` seems to fail!
